browser/brave_ads/device_id/device_id_impl.cc

```diff
@@ -20,20 +22,16 @@ namespace brave_ads {
 
 namespace {
 
-bool ComputeHmacSha256(const std::string& key,
-                       const std::string& text,
-                       std::string* signature_return) {
+std::optional<std::string> ComputeHmacSha256(std::string_view key,
+                                             std::string_view text) {
   crypto::HMAC hmac(crypto::HMAC::SHA256);
-  const size_t digest_length = hmac.DigestLength();
-  std::vector<uint8_t> digest(digest_length);
-  const bool result =
-      hmac.Init(key) && hmac.Sign(text, &digest[0], digest.size());
-  if (result) {
-    *signature_return =
-        base::ToLowerASCII(base::HexEncode(digest.data(), digest.size()));
+  std::array<uint8_t, 32u> digest;
```

Comment:
It's a shame that we cannot get digest length in compile time. Can we add some check that `hmac.DigestLength()` is equal to array size?

Reply:
I did consider that, but then I checked upstream, and how they are using this object, and I decided against. Here's what `DigestLength` looks like:

```c++
size_t HMAC::DigestLength() const {
  switch (hash_alg_) {
    case SHA1:
      return 20;
    case SHA256:
      return 32;
    default:
      NOTREACHED();
  }
}
```

I've checked many places where the key is created, and when it is created on the stack they never check:

```c++
  crypto::HMAC response(crypto::HMAC::SHA256);
  if (!response.Init(tag)) { return; }
  unsigned char out_bytes[kSharedSecretHashLength];

  // ...
 return std::string(out_bytes, out_bytes + sizeof(out_bytes));
```

and 

```c++
  auto reply = fido_parsing_utils::Materialize(hello_msg);
  crypto::HMAC hmac(crypto::HMAC::SHA256);
  if (!hmac.Init(handshake_key))
    return std::vector<uint8_t>();

  std::array<uint8_t, 32> authenticator_hello_mac;
  if (!hmac.Sign(fido_parsing_utils::ConvertToStringView(hello_msg),
                 authenticator_hello_mac.data(),
                 authenticator_hello_mac.size())) {
    return std::vector<uint8_t>();
  }
```

The only place they check is wherever they are using `vector`. I feel a check would add noise here to no benefit, as this is part of the contract of the type.

Reply:
Also, I just realised we don't need the check because we are passing digest as a `span`, and it gets consumed like this:

```
bool HMAC::Sign(base::span<const uint8_t> data,
                base::span<uint8_t> digest) const {
  DCHECK(initialized_);

  if (digest.size() > DigestLength())
    return false;

  ScopedOpenSSLSafeSizeBuffer<EVP_MAX_MD_SIZE> result(digest.data(),
                                                      digest.size());
  return !!::HMAC(hash_alg_ == SHA1 ? EVP_sha1() : EVP_sha256(), key_.data(),
                  key_.size(), data.data(), data.size(), result.safe_buffer(),
                  nullptr);
}
```